### PR TITLE
*: add data type default values 

### PR DESCRIPTION
**Pull request by hfxsd**
### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions](https://github.com/pingcap/docs/blob/master/CONTRIBUTING.md#guideline-for-choosing-the-affected-versions).

- [x] master (the latest development version)
- [x] v8.0 (TiDB 8.0 versions)
- [ ] v7.6 (TiDB 7.6 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.4 (TiDB 7.4 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:https://github.com/pingcap/docs-cn/pull/16751
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: data-type-default-values.md [hunk 1]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
 - For date and time types other than `TIMESTAMP`, the default is the appropriate "zero" value for the type. For `TIMESTAMP`, the default value is the current date and time.
 - For string types other than `ENUM`, the default value is the empty string. For `ENUM`, the default is the first enumeration value.
+
+## Specifying expressions as default values
+
+> **Warning:**
+>
+> - Currently, this feature is experimental. It is not recommended that you use it in production environments. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
+
+Starting from v8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit Default Handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying expressions as default values in the `DEFAULT` clause.
```
Comment (Frank945946) [inline data-type-default-values.md:35]:
```suggestion
Starting from v8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit Default Handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause.
```

**Review by Frank945946 — APPROVED** [review]

File: data-type-default-values.md [hunk 2]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+CREATE TABLE t2 (b BLOB DEFAULT (rand()));
+```
+
+TiDB currently supports the following expressions:
+
+* `RAND`, `UUID`, `UUID_TO_BIN`
+
+Starting from TiDB v8.0.0, the `DEFAULT` clause supports using the following expressions to set default values.
+
+* `UPPER(SUBSTRING_INDEX(user(),'@',1))`
```
Comment (hfxsd) [inline data-type-default-values.md:49]:
```suggestion
* `UPPER(SUBSTRING_INDEX(user(), '@', 1))`
```

File: data-type-default-values.md [hunk 3]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+* `REPLACE(UPPER(uuid()), '-', '')`
+
+* The `DATE_FORMAT` supports the following formats:
+
+    * `DATE_FORMAT(NOW(),'%Y-%m')`
+    * `DATE_FORMAT(NOW(),'%Y-%m-%d')`
+    * `DATE_FORMAT(NOW(),'%Y-%m-%d %H.%i.%s')`
+    * `DATE_FORMAT(NOW(),'%Y-%m-%d %H:%i:%s')`
+
+* `STR_TO_DATE('1980-01-01','%Y-%m-%d')`
```
Comment (hfxsd) [inline data-type-default-values.md:60]:
```suggestion
    * `DATE_FORMAT(NOW(), '%Y-%m')`
    * `DATE_FORMAT(NOW(), '%Y-%m-%d')`
    * `DATE_FORMAT(NOW(), '%Y-%m-%d %H.%i.%s')`
    * `DATE_FORMAT(NOW(), '%Y-%m-%d %H:%i:%s')`

* `STR_TO_DATE('1980-01-01', '%Y-%m-%d')`
```

**Review by tangenta — APPROVED** [review]

File: data-type-default-values.md [hunk 4]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+    * `DATE_FORMAT(NOW(), '%Y-%m')`
+    * `DATE_FORMAT(NOW(), '%Y-%m-%d')`
+    * `DATE_FORMAT(NOW(), '%Y-%m-%d %H.%i.%s')`
+    * `DATE_FORMAT(NOW(), '%Y-%m-%d %H:%i:%s')`
+
+* `STR_TO_DATE('1980-01-01', '%Y-%m-%d')`
+
+> **Note:**
+>
+> Currently, the `CHANGE COLUMN` and `MODIFY COLUMN` statements do not support using expressions as default values.
```
Comment (hfxsd) [inline data-type-default-values.md:64]:
```suggestion
> Currently, the `ADD COLUMN` statement does not support using expressions as default values.
```

File: data-type-default-values.md [hunk 5]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+Starting from TiDB v8.0.0, the `BLOB`, `TEXT`, and `JSON` data types support default values, but the default values can only be set as expressions.
+
+```sql
+CREATE TABLE t2 (b BLOB DEFAULT (rand()));
+```
+
+TiDB currently supports the following expressions:
+
+* `RAND`, `UUID`, `UUID_TO_BIN`
```
Comment (hfxsd) [inline data-type-default-values.md:45]:
```suggestion
* `RAND`, `UUID`, `UUID_TO_BIN`
```

File: data-type-default-values.md [hunk 6]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+Starting from TiDB v8.0.0, the `DEFAULT` clause supports using the following expressions to set default values.
+
+* `UPPER(SUBSTRING_INDEX(user(), '@', 1))`
```
Comment (dveeden) [inline data-type-default-values.md:49]:
```suggestion
* `UPPER(SUBSTRING_INDEX(USER(), '@', 1))`
```

File: data-type-default-values.md [hunk 7]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+* `REPLACE(UPPER(uuid()), '-', '')`
```
Comment (dveeden) [inline data-type-default-values.md:51]:
```suggestion
* `REPLACE(UPPER(UUID()), '-', '')`
```

File: data-type-default-values.md [hunk 8]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+## Specify expressions as default values
+
+> **Warning:**
+>
+> - Currently, this feature is experimental. It is not recommended that you use it in production environments. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
+
+Starting from v8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit Default Handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause.
+
+Starting from TiDB v8.0.0, the `BLOB`, `TEXT`, and `JSON` data types support default values, but the default values can only be set as expressions.
```
Comment (dveeden) [inline data-type-default-values.md:37]:
```suggestion
Starting from TiDB v8.0.0, the [`BLOB`, `TEXT`](/data-type-string.md), and [`JSON`](/data-type-json.md) data types support default values, but the default values can only be set as expressions.
```

**Comment by dveeden** [comment]
Do we need to add anything related to https://github.com/pingcap/tidb/issues/49508 ? 

And https://github.com/pingcap/tidb/issues/45506 seems to suggest that more functions might be supported.

**Comment by dveeden** [comment]
Looking at [`getFuncCallDefaultValue()`](https://github.com/pingcap/tidb/blob/0878b1518afc9354ddc08a94381bcfd34bd7f24b/pkg/ddl/ddl_api.go#L1286) it looks like these are supported:

- `CURRENT_TIMESTAMP()`,
- `CURRENT_DATE()`
- `NEXTVAL()`
- `RAND()`
- `UUID()`
- `UUID_TO_BIN()`
- `DATE_FORMAT()`
- `REPLACE()` 
- `UPPER()`
- `STR_TO_DATE()`

Some of these (`REPLACE()`, `UPPER()`, ...) are only supported with specific arguments.

**Review by zimulala — APPROVED** [review]

**Review by dveeden — APPROVED** [review]

**Comment by zimulala** [comment]
> Looking at [`getFuncCallDefaultValue()`](https://github.com/pingcap/tidb/blob/0878b1518afc9354ddc08a94381bcfd34bd7f24b/pkg/ddl/ddl_api.go#L1286) it looks like these are supported:
> 
> * `CURRENT_TIMESTAMP()`,
> * `CURRENT_DATE()`
> * `NEXTVAL()`
> * `RAND()`
> * `UUID()`
> * `UUID_TO_BIN()`
> * `DATE_FORMAT()`
> * `REPLACE()`
> * `UPPER()`
> * `STR_TO_DATE()`
> 
> Some of these (`REPLACE()`, `UPPER()`, ...) are only supported with specific arguments.

* `CURRENT_DATE` and `CURRENT_DATE` already supported on line9.
* `REPLACE`, `UPPER`, `STR_TO_DATE` supported as this PR, and the related issue: https://github.com/pingcap/tidb/issues/50936

File: data-type-default-values.md [hunk 9]
```diff
@@ -8,7 +8,7 @@ aliases: ['/docs/dev/data-type-default-values/','/docs/dev/reference/sql/data-ty
 
 The `DEFAULT` value clause in a data type specification indicates a default value for a column. The default value must be a constant and cannot be a function or an expression. But for the time type, you can specify the `NOW`, `CURRENT_TIMESTAMP`, `LOCALTIME`, and `LOCALTIMESTAMP` functions as the default for `TIMESTAMP` and `DATETIME` columns.
 
-The `BLOB`, `TEXT`, and `JSON` columns __cannot__ be assigned a default value.
+The `BLOB`, `TEXT`, and `JSON` columns can be assigned a default value.
```
Comment (Oreoxmt) [inline data-type-default-values.md:11]:
```suggestion
Starting from v8.0.0, TiDB supports [specifying expressions as default values](#specify-expressions-as-default-values) for [`BLOB`](/data-type-string.md#blob-type), [`TEXT`](/data-type-string.md#text-type), and [`JSON`](/data-type-json.md#json-type) data types.
```

File: data-type-default-values.md [hunk 10]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
 - For numeric types, the default is 0. If declared with the `AUTO_INCREMENT` attribute, the default is the next value in the sequence.
 - For date and time types other than `TIMESTAMP`, the default is the appropriate "zero" value for the type. For `TIMESTAMP`, the default value is the current date and time.
 - For string types other than `ENUM`, the default value is the empty string. For `ENUM`, the default is the first enumeration value.
+
+## Specify expressions as default values
+
+> **Warning:**
+>
+> - Currently, this feature is experimental. It is not recommended that you use it in production environments. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
```
Comment (Oreoxmt) [inline data-type-default-values.md:33]:
```suggestion
> Currently, this feature is experimental. It is not recommended that you use it in production environments. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
```

File: data-type-default-values.md [hunk 11]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+Starting from v8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit Default Handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause.
```
Comment (Oreoxmt) [inline data-type-default-values.md:35]:
```suggestion
Starting from 8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit default handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause.
```

File: data-type-default-values.md [hunk 12]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+Starting from TiDB v8.0.0, the [`BLOB`, `TEXT`](/data-type-string.md), and [`JSON`](/data-type-json.md) data types support default values, but the default values can only be set as expressions.
```
Comment (Oreoxmt) [inline data-type-default-values.md:37]:
```suggestion
Starting from v8.0.0, TiDB supports assigning default values to `BLOB`, `TEXT`, and `JSON` data types. However, these default values can only be set as expressions.
```

File: data-type-default-values.md [hunk 13]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+## Specify expressions as default values
+
+> **Warning:**
+>
+> Currently, this feature is experimental. It is not recommended that you use it in production environments. This feature might be changed or removed without prior notice. If you find a bug, you can report an [issue](https://github.com/pingcap/tidb/issues) on GitHub.
+
+Starting from 8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit default handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause.
+
+Starting from v8.0.0, TiDB supports assigning default values to `BLOB`, `TEXT`, and `JSON` data types. However, these default values can only be set as expressions.
```
Comment (hfxsd) [inline data-type-default-values.md:37]:
```suggestion
Starting from 8.0.13, MySQL supports specifying expressions as default values in the `DEFAULT` clause. For more information, see [Explicit default handling as of MySQL 8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults-explicit). 

TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause. Starting from v8.0.0, TiDB supports assigning default values to `BLOB`, `TEXT`, and `JSON` data types. However, these default values can only be set as expressions. The following is an example of `BLOB`:
```

**Review by Oreoxmt — APPROVED** [review]

File: data-type-default-values.md [hunk 14]
```diff
@@ -25,3 +25,40 @@ Implicit defaults are defined as follows:
...
+
+TiDB has implemented this feature and supports specifying some expressions as default values in the `DEFAULT` clause. Starting from v8.0.0, TiDB supports assigning default values to `BLOB`, `TEXT`, and `JSON` data types. However, these default values can only be set as expressions. The following is an example of `BLOB`:
+
+```sql
+CREATE TABLE t2 (b BLOB DEFAULT (RAND()));
+```
+
+TiDB currently supports the following expressions:
+
+* [`RAND()`](/functions-and-operators/numeric-functions-and-operators.md), [`UUID()`](/functions-and-operators/miscellaneous-functions.md), [`UUID_TO_BIN()`](/functions-and-operators/miscellaneous-functions.md)
```
Comment (hfxsd) [inline data-type-default-values.md:45]:
```suggestion
* [`RAND()`](/functions-and-operators/numeric-functions-and-operators.md)
* [`UUID()`](/functions-and-operators/miscellaneous-functions.md)
* [`UUID_TO_BIN()`](/functions-and-operators/miscellaneous-functions.md)
```

**Pull request merged**